blargh.. of course its a problem on esp8266
```
/Users/bdraco/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: /Users/bdraco/.platformio/packages/framework-arduinoespressif8266/tools/sdk/lib/libstdc++.a(new_opa.o):(.literal._ZnwjSt11align_val_t+0x4): undefined reference to `memalign'
/Users/bdraco/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: /Users/bdraco/.platformio/packages/framework-arduinoespressif8266/tools/sdk/lib/libstdc++.a(new_opa.o): in function `aligned_alloc':
/workdir/repo/gcc-gnu/libstdc++-v3/libsupc++/new_opa.cc:86: undefined reference to `memalign'
```

--- a/esphome/components/sha256/sha256.cpp
+++ b/esphome/components/sha256/sha256.cpp
@@ -10,11 +10,11 @@ namespace esphome::sha256 {
 
 #if defined(USE_ESP32) || defined(USE_LIBRETINY)
 
-// CRITICAL ESP32-S2/S3 HARDWARE SHA ACCELERATION REQUIREMENTS (IDF 5.5.x):
+// CRITICAL ESP32 HARDWARE SHA ACCELERATION REQUIREMENTS (IDF 5.5.x):
 //
-// The ESP32-S2/S3 uses hardware DMA for SHA acceleration. The DMA engine requires proper
-// alignment of the digest output buffer. This is handled automatically via HashBase::digest_
-// which has alignas(32). This imposes two critical constraints:
+// ESP32 variants (except original ESP32) use DMA-based hardware SHA acceleration that requires
+// 32-byte aligned digest buffers. This is handled automatically via HashBase::digest_ which has
+// alignas(32) on these platforms. Two additional constraints apply:
 //
 // 1. NO VARIABLE LENGTH ARRAYS (VLAs): VLAs corrupt the stack layout, causing the DMA engine to
 //    write to incorrect memory locations. This results in null pointer dereferences and crashes.
@@ -35,7 +35,7 @@ namespace esphome::sha256 {
 //     // hasher destroyed when function returns
 //   }
 //
-// INCORRECT USAGE (WILL FAIL ON ESP32-S2/S3):
+// INCORRECT USAGE (WILL FAIL):
 //   void my_function() {
 //     sha256::SHA256 hasher;
 //     helper(&hasher);  // WRONG: Passed to different stack frame

--- a/esphome/components/sha256/sha256.h
+++ b/esphome/components/sha256/sha256.h
@@ -24,7 +24,7 @@ namespace esphome::sha256 {
 
 /// SHA256 hash implementation.
 ///
-/// CRITICAL for ESP32-S2/S3 with IDF 5.5.x hardware SHA acceleration:
+/// CRITICAL for ESP32 variants (except original) with IDF 5.5.x hardware SHA acceleration:
 /// 1. The object MUST stay in the same stack frame (no passing to other functions)
 /// 2. NO Variable Length Arrays (VLAs) in the same function
 ///

--- a/esphome/core/hash_base.h
+++ b/esphome/core/hash_base.h
@@ -44,9 +44,10 @@ class HashBase {
   virtual size_t get_size() const = 0;
 
  protected:
-// ESP32-S2/S3 hardware SHA uses DMA that requires 32-byte aligned buffers.
-// Other platforms either don't have hardware SHA or don't require alignment.
-#if defined(USE_ESP32_VARIANT_ESP32S2) || defined(USE_ESP32_VARIANT_ESP32S3)
+// ESP32 variants with DMA-based hardware SHA (all except original ESP32) require 32-byte aligned buffers.
+// Original ESP32 uses a different hardware SHA implementation without DMA alignment requirements.
+// Other platforms (ESP8266, RP2040, LibreTiny) use software SHA and don't need alignment.
+#if defined(USE_ESP32) && !defined(USE_ESP32_VARIANT_ESP32)
   alignas(32)
 #endif
       uint8_t digest_[32];  // Storage sized for max(MD5=16, SHA256=32) bytes

--- a/esphome/core/hash_base.h
+++ b/esphome/core/hash_base.h
@@ -44,9 +44,12 @@ class HashBase {
   virtual size_t get_size() const = 0;
 
  protected:
-  // 32-byte alignment required for ESP32-S2/S3 hardware SHA DMA operations.
-  // This also sets the class alignment to 32, ensuring derived objects are properly aligned.
-  alignas(32) uint8_t digest_[32];  // Storage sized for max(MD5=16, SHA256=32) bytes
+// ESP32-S2/S3 hardware SHA uses DMA that requires 32-byte aligned buffers.
+// Other platforms either don't have hardware SHA or don't require alignment.
+#if defined(USE_ESP32_VARIANT_ESP32S2) || defined(USE_ESP32_VARIANT_ESP32S3)
+  alignas(32)
+#endif
+      uint8_t digest_[32];  // Storage sized for max(MD5=16, SHA256=32) bytes
 };
 
 }  // namespace esphome

